firestore/src/Connector.gs

```diff
@@ -0,0 +1,288 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/** 
+ * Contains core functionality for Data Studio connector. 
+ *
+ * @constructor
+ */
+function Connector(enableLogging) {
+
+  /** @const */
+  this.FIELD_TYPES = [
+    'STRING',
+    'NUMBER',
+    'BOOLEAN',
+    'TIMESTAMP'
+  ];
+
+  /** @const */
+  this.TIMESTAMP_SEMANTICS = {
+    conceptType: 'DIMENSION',
+    semanticGroup: 'DATE_AND_TIME',
+    semanticType: 'YEAR_MONTH_DAY_HOUR'
+  };
+  
+  /** @const */
+  this.cloud = new GoogleCloud();
+  
+  /** @const */
+  this.enableLogging = enableLogging;
+}
+
+
+/**
+ * Stringifies parameters and responses for a given function and logs them to
+ * Stackdriver.
+ *
+ * @param {string} functionName Function to be logged and executed.
+ * @param {Object} parameter Parameter for the `functionName` function.
+ * @returns {any} Returns the response of `functionName` function.
+ */
+Connector.prototype.logAndExecute = function(functionName, parameter) {
+  if (this.enableLogging) {
+    const paramString = JSON.stringify(parameter, null, 2);
+    console.log([functionName, 'request', paramString]);
+  }
+
+  const returnObject = this[functionName](parameter);
+
+  if (this.enableLogging) {
+    const returnString = JSON.stringify(returnObject, null, 2);
+    console.log([functionName, 'response', returnString]);
+  }
+
+  return returnObject;
+};
+
+
+/** 
+ * Predefined Data Studio function for specifying connector configuration.
+ *
+ * @param {Object} request Config request parameters.
+ * @return {Object} Connector configuration to be displayed to the user.
+ */
+Connector.prototype.getConfig = function(request) {
+  const projects = this.cloud.listCloudProjects();
+  const options = [];
+  projects.forEach(function(project) {
+    options.push({
+      label: project.name + ' (' + project.projectId + ')',
+      value: project.projectId
+    });
+  });
+  const optionsSorted = options.sort(function(x, y) {
+    const xId = x.label.toLowerCase();
+    const yId = y.label.toLowerCase();
+    return xId == yId ? 0 : (xId < yId ? -1 : 1);
+  });
+    
+  return {
+    configParams: [
+      {
+        name: 'project',
+        type: 'SELECT_SINGLE',
+        displayName: 'Firebase Project',
+        helpText: 'Select a Google Cloud/Firebase project that uses Firestore.',
+        options: options
+      },
+      {
+        name: 'collection',
+        type: 'TEXTINPUT',
+        displayName: 'Collection',
+        helpText: 'Select the Firestore collection to use for this Data Source. To use multiple \
+                   collections, create additional Data Sources.'
```

Comment:
I think this creates a string with a bunch of whitespace between `multiple` and `collections`. Could you double check this behavior and add a replaceAll, if necessary so things are spaced as expected. It might display correctly in DS as is, but that's probably an implementation detail.

Reply:
Here and throughout, of course.

Reply:
Confirmed that spacing works as-is.